otri/downloader/__init__.py

```diff
@@ -808,3 +814,420 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
                 Anything that the caller function can pass.\n
         '''
         return atom
+
+
+class AdapterComponent(ABC):
+    '''
+    An adapter component is an object used by an adapter to perform standard operation between adapters of different sources.
+    A component implements some of its interface's methods, the less the better.
+    '''
+
+    def prepare(self, **kwargs) -> Mapping:
+        '''
+        First of the download functionalities.
+        '''
+        return kwargs
+
+    def retrieve(self, data_stream: WritableStream, **kwargs):
+        '''
+        Second of the download functionalities.
+
+        Parameters:
+            data_stream : WritableStream
+                Stream where to place downloaded data do be parsed by an atomizer function.
+        '''
+        return kwargs
+
+    def atomize(self, data_stream: ReadableStream, output_stream: WritableStream, **kwargs):
+        '''
+        Third of the download functionalities.
+
+        Parameters:
+            data_stream : ReadableStream
+                Stream where downloaded data is placed to be parsed.
+            output_stream : WritableStream
+                Stream where to place parsed data to be read by others.
+        '''
+        return kwargs
+
+
+class Adapter(ABC):
+    '''
+    Imports data from an external source.
+    '''
+
+    def __init__(self, components: List[AdapterComponent] = list(), sync : bool = True):
+        '''
+        Initialises adapter.
+
+        Parameters:
+            components : list[AdapterComponent]
+                Ordered list of adapter components that will be called on download. Default empty list.
```

Comment:
Specify that the components will be called in the given order.

Reply:
Also, why not using variables like django does insead of passing parameters to constructor? The parameters are static for every class.

Reply:
I did indeed end up having components as class attributes instead of constructor, it looks incredibly cleaner.
It cannot be applied to adapters too as it's more comfortable passing parameters in the constructor rather then writing the whole class just to set the parameters.